File: cli_ure/source/uno_bridge/cli_environment.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */

#include "Cli_environment.h"

#using <cli_ure.dll>
#using <system.dll>
#include "osl/diagnose.h"
#include "cli_proxy.h"

using namespace System::Runtime::Remoting;
using namespace System::Runtime::Remoting::Proxies;
using namespace System::Collections;
using namespace System::Text;
using namespace System::Diagnostics;
using namespace System::Threading;

namespace cli_uno
{

inline System::String^ Cli_environment::createKey(System::String^ oid, System::Type^ t)
{
    return System::String::Concat(oid, t->FullName);
}


inline Cli_environment::Cli_environment()
{
#if OSL_DEBUG_LEVEL >= 2
    _numRegisteredObjects = 0;
#endif
}

Cli_environment::~Cli_environment()
{
    OSL_ENSURE(_numRegisteredObjects == 0,
               "cli uno bridge: CLI environment contains unrevoked objects");
}


System::Object^ Cli_environment::registerInterface(
    System::Object^ obj, System::String^ oid)
{
#if OSL_DEBUG_LEVEL >= 1
    //obj must be a transparent proxy
    OSL_ASSERT(RemotingServices::IsTransparentProxy(obj));
    _numRegisteredObjects ++;
#endif
    OSL_ASSERT( ! m_objects->ContainsKey(oid));
    m_objects->Add(oid, gcnew WeakReference(obj));
    return obj;
}
System::Object^ Cli_environment::registerInterface      (
    System::Object^ obj, System::String^ oid, System::Type^ type)
{
#if OSL_DEBUG_LEVEL >= 1
    //obj must be a real cli object
    OSL_ASSERT( ! RemotingServices::IsTransparentProxy(obj));
    _numRegisteredObjects ++;
#endif
    System::String^ key = createKey(oid, type);
    //see synchronization in map_uno2cli in cli_data.cxx
    OSL_ASSERT( ! m_objects->ContainsKey(key));
    m_objects->Add(key, gcnew WeakReference(obj));
    return obj;
}

void Cli_environment::revokeInterface(System::String^ oid, System::Type^ type)
{
    System::String^ key = type != nullptr ? createKey(oid, type) : oid;
#if OSL_DEBUG_LEVEL >= 1
    _numRegisteredObjects --;
#endif
#if OSL_DEBUG_LEVEL >= 2
    int i = 1;
    if (m_objects->ContainsKey(key) == false)
    {
        Trace::WriteLine("cli uno bridge: try to revoke unregistered interface");
        Trace::WriteLine(oid);
        i = 0;
    }
    Trace::WriteLine(System::String::Format(
                         new System::String(S"cli uno bridge: {0} remaining registered interfaces"),
                         __box(m_objects->get_Count() - 1)));
#endif
    m_objects->Remove(key);
}

inline void Cli_environment::revokeInterface(System::String^ oid)
{
    return revokeInterface(oid, nullptr);
}

System::Object^ Cli_environment::getRegisteredInterface(System::String^ oid,
                                                        System::Type^ type)
{
    //try if it is a UNO interface
    System::Object^ ret = nullptr;
    ret = m_objects[oid];
    if (! ret)
    {
        //try if if it is a proxy for a cli object
        oid = createKey(oid, type);
        ret = m_objects[ oid ];
    }
    if (ret != nullptr)
    {
        System::WeakReference^ weakIface =
            static_cast< System::WeakReference ^ >( ret );
        ret = weakIface->Target;
    }
    if (ret == nullptr)
        m_objects->Remove( oid );
    return ret;
}

System::String^ Cli_environment::getObjectIdentifier(System::Object^ obj)
{
    System::String^ oId= nullptr;
    RealProxy^ aProxy= RemotingServices::GetRealProxy(obj);
    if (aProxy)
    {
        UnoInterfaceProxy^ proxyImpl= dynamic_cast<UnoInterfaceProxy^>(aProxy);
        if (proxyImpl)
            oId= proxyImpl->getOid();
    }

    if (oId == nullptr)
    {
        StringBuilder ^ buf= gcnew StringBuilder(256);
        bool bFirst = new bool(false);
        System::Threading::Monitor::Enter(Cli_environment::typeid);
        try {
            buf->Append(m_IDGen->GetId(obj, bFirst));
        } finally
        {
            System::Threading::Monitor::Exit(Cli_environment::typeid);
        }

        buf->Append(sOidPart);
        oId= buf->ToString();
    }
    return oId;
}
} //namespace cli_uno

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */

File: cli_ure/source/mono_bridge/cli_environment.cs
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*************************************************************************
 *
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 * 
 * Copyright 2000, 2010 Oracle and/or its affiliates.
 *
 * OpenOffice.org - a multi-platform office productivity suite
 *
 * This file is part of OpenOffice.org.
 *
 * OpenOffice.org is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License version 3
 * only, as published by the Free Software Foundation.
 *
 * OpenOffice.org is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License version 3 for more details
 * (a copy is included in the LICENSE file that accompanied this code).
 *
 * You should have received a copy of the GNU Lesser General Public License
 * version 3 along with OpenOffice.org.  If not, see
 * <http://www.openoffice.org/license.html>
 * for a copy of the LGPLv3 License.
 *
 ************************************************************************/

namespace cli_uno
{

using System;
using System.Collections;
using System.Runtime.Remoting;
using System.Runtime.Remoting.Proxies;
using System.Text;

using com.sun.star.bridges.mono_uno;

public class Cli_environment
{
    static string sOidPart = ";cli[0];";

    static Hashtable m_objects = Hashtable.Synchronized(new Hashtable());

    static string createKey(string oid, Type t)
    {
        return oid + t.FullName;
    }

//     FIXME setup debugging info here
//     public Cli_environment()
//     {
//     }

//     FIXME assert there are no more registered objects
//     public ~Cli_environment()
//     {
//     }

    /**
       Registers an UNO object as being mapped by this bridge. The resulting
       cli object is represents all interfaces of the UNO object. Therefore the
       object can be registered only with its OID; a type is not necessary.
    */
    public object registerInterface(object obj, string oid)
    {
        // FIXME debugging stuff
        m_objects.Add(oid, obj); // new WeakReference(obj));
        return obj;
    }

    /**
       Registers a CLI object as being mapped by this bridge. The resulting
       object represents exactly one UNO interface.
     */
    public object registerInterface(object obj, string oid, Type type)
    {
        // FIXME debugging stuff
        string key = createKey(oid, type);
        m_objects.Add(key, obj); // new WeakReference(obj));
        return obj;
    }

    /**
       By revoking an interface it is declared that the respective interface has
       not been mapped. The proxy implementations call revoke interface in their
       destructors.
     */
    public void revokeInterface(string oid)
    {
        revokeInterface(oid, null);
    }

    public void revokeInterface(string oid, Type type)
    {
        // FIXME debugging stuff
        string key = type != null ? createKey(oid, type) : oid;
        m_objects.Remove(key);
    }

    /**
     * Retrieves an interface identified by its object id and type from this
     * environment.
     *
     * @param oid object id of interface to be retrieved
     * @param type the type description of the interface to be retrieved
     * @see com.sun.star.uno.IEnvironment#getRegisteredInterface
     */
    public object getRegisteredInterface(string oid, Type type)
    {
        // try if it is a UNO interface
        object ret = null;
        ret = m_objects[oid];
        if (ret == null)
        {
            // try if it is a proxy for a cli object
            oid = createKey(oid, type);
            ret = m_objects[oid];
        }
/*        if (ret != null)
        {
            WeakReference weakIface = (WeakReference)ret;
            ret = weakIface.Target;
        } */

        if (ret == null)
            m_objects.Remove(oid);

        return ret;
    }

        /**
         * Generates a worldwide unique object identifier (oid) for the given object. It is
     * guaranteed, that subsequent calls to the method with the same object
     * will give the same id.
         * <p>
         * @return     the generated oid.
         * @param      object     the object for which a Oid should be generated.
         */
    public static string getObjectIdentifier(object obj)
    {
        string oid = null;
        RealProxy realProxy = null;

        if (RemotingServices.IsTransparentProxy(obj))
            realProxy = RemotingServices.GetRealProxy(obj);

        if (realProxy != null)
        {
            UnoInterfaceProxy proxyImpl = realProxy as UnoInterfaceProxy;
            if (proxyImpl != null)
                oid = proxyImpl.Oid;
        }

        if (oid == null)
        {
            Guid gd = typeof(Cli_environment).GUID; // FIXME apparently not a good idea with mono
            StringBuilder buf = new StringBuilder(128);
            buf.Append(obj.GetHashCode());
            buf.Append(sOidPart);
            buf.Append(gd);
            oid = buf.ToString();
        }

        return oid;
    }

}

}
